indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceRecordReader.java

```diff
@@ -0,0 +1,168 @@
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.io.Files;
+import com.metamx.common.ISE;
+import io.druid.data.input.MapBasedInputRow;
+import io.druid.data.input.MapBasedRow;
+import io.druid.query.filter.DimFilter;
+import io.druid.segment.IndexIO;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.QueryableIndexStorageAdapter;
+import io.druid.segment.StorageAdapter;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.realtime.firehose.IngestSegmentFirehose;
+import io.druid.segment.serde.ComplexMetricSerde;
+import io.druid.segment.serde.ComplexMetrics;
+import io.druid.utils.CompressionUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class DatasourceRecordReader extends RecordReader<NullWritable, HadoopWritableInputRow>
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(DatasourceRecordReader.class);
+
+  private SegmentLoadSpec spec;
+  private Iterable<Metric> complexMetrics;
+
+  private IngestSegmentFirehose rowYielder;
+
+  private MapBasedRow currRow;
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException
+  {
+    spec = readSegmentJobSpec(context.getConfiguration(), new ObjectMapper());
+
+    final List<String> dimensions = spec.getDimensions();
+    final List<String> metrics = Lists.transform(
+        spec.getMetrics(),
+        new Function<Metric, String>()
+        {
+          @Nullable
+          @Override
+          public String apply(Metric input)
+          {
+            return input.getName();
+          }
+        }
+    );
+
+    complexMetrics = Iterables.filter(
+        spec.getMetrics(), new Predicate<Metric>()
+        {
+          @Override
+          public boolean apply(@Nullable Metric input)
+          {
+            return input.getType() == ValueType.COMPLEX;
+          }
+        }
+    );
+    final DimFilter filter = spec.getFilter();
+
+    String hdfsPath = ((DatasourceInputSplit) split).getPath();
+    logger.info("Reading segment from " + hdfsPath);
+
+    File segmentDir = Files.createTempDir();
+    logger.info("segment dir: " + segmentDir);
+
+    FileSystem fs = FileSystem.get(context.getConfiguration());
+    getSegmentFiles(hdfsPath, segmentDir, fs);
+    logger.info("finished getting segment files");
+
+    QueryableIndex index = IndexIO.loadIndex(segmentDir);
```

Comment:
For now `QueryableIndex` implements closeable and should have `close()` called on it